src/datetime.rs

```diff
@@ -235,6 +235,43 @@ impl<Tz: TimeZone> DateTime<Tz> {
     }
 }
 
+impl From<DateTime<Utc>> for DateTime<FixedOffset> {
+    fn from(src: DateTime<Utc>) -> Self {
```

Comment:
> ... actually, probably there shouldn't be a From impl for FixedOffset in this first pass, since there is no way to specify what the offset should be.

Yup, I considered that but ultimately came to a different conclusion. The problem is that `FixedOffset` is a type but the offset itself is independent of the type. Let's say you read/parse/have a `DateTime<Utc>` but want to assign its value to a type that is normally `FixedOffset`. Apart from their having different offsets they're also completely different types to rust's type system, and you can't assign from one to the other at all. However, since it's possible to assign from a `DateTime<FixedOffset>` with an offset of -1 to a `DateTime<FixedOffset>` with a fixed offset of +2, it should also be possible to assign a `DateTime<Utc>` object. 

Honestly the type split is very painful and the fact that a `DateTime<Utc>` is not the same type as `DateTime<FixedOffset>` but two different fixed offsets are the same type isn't very rational, however with the complexity of dates and times I can definitely see why it was done that way. However, I do think it makes more sense to just unify it all with a `DateTime<Offset>` type where `Offset` is an enum with variants `Local, Utc, Fixed(offset: i32)`. That way they all have the same basic type (since given `A<B> where B: enum` two objects `A<B::Variant1>` and `A<B::Variant2>` have the same type in rust's type system), and that takes care of one of the issues altogether.

Reply:
Two responses:

* You've convinced me about `From<Utc>` for `FixedOffset`, assuming that we put some documentation in pointing at `with_timezone` for people who want a specific offset.

In terms of enums vs generics, the downside to that is that it requires a switch on enums every time that you want to interact with it, and the size of the time is going to be larger (containing the enum discriminant). The way that it is right now, everyone using Utc get's zero-sized static method dispatch, which is neat, at least. It is a little less ergonomic, but it also prevents certain classes of errors. In general I'm of the philosophy that people shouldn't be mixing their timezones most of the time, although I admit that there are probably use cases that I haven't considered.

Reply:
@mqudsi are you interested in finishing this, or should I take over?

Reply:
Sorry, yes, I should be able to push out the requested changes.

WRT your earlier reply: I didn't consider the zero overhead size, and yes, that is nice bonus to the existing approach. Since RFC 2000 was merged, when support for (const generics)[https://internals.rust-lang.org/t/lang-team-minutes-const-generics/5090] lands it should actually be possible to include the timezone offset in the type itself, allowing for zero-cost generics to convert between timezones and enforce underlying timezone offsets at build-time.